lab2/README.md

```diff
@@ -0,0 +1,648 @@
+- Deferring tasks to softirqs allows the interrupt handler to return quickly, minimizing the time the CPU is unavailable for other tasks.
+
+### L2
+#### queue
+> Maintained by kernel
+* Each queue will contain a pointer to the associated device and a pointer to the ingress/egress sk_buff.
+* The loopback device is special as it does not require a queue at all.
+
+#### What Handler(Device Driver) Do?
+1. Copy the frame into the sk_buff structure.
+2. Initialize the sk_buff settings for higher-level handlers.
+    For example, set skb→protocol to specify the upper-layer protocol.
+3. Schedule the NET_RX_SOFTIRQ softirq to notify the kernel of the arrival of a new frame.
+    Also, place the frame in the CPU's private queue (stored in each CPU's separate ```softnet_data```).
+
+#### How to Notify Kernel
+* Old Method: netif_rx
+    * Some device drivers use this method.
+    *  Typically called in an interrupt context, it may temporarily disable CPU interrupts.
+* NAPI (New API)
+    * The main idea is to combine interrupt and polling mechanisms.
+    * New interrupts are not generated if the kernel has not finished processing other packets in the queue.
+    * Reduces CPU loading during high workloads by decreasing the number of interrupts.
+#### Processing Frame (Receive)
+[```netif_receive_skb```](<https://github.com/torvalds/linux/blob/master/net/core/dev.c>)
+
+Each driver is associated with a specific hardware type (e.g., Ethernet), allowing it to interpret the L2 header and identify the L3 protocol.
+
+Main Functions:
+
+1. Pass the frame copy to the protocol demultiplexer.
+2. Forward the frame copy to the L3 handler specified by skb→protocol.
+3. Handle layer-specific functions, such as bridging.
+#### Transmission Enable & Disable:
+
+When a device driver detects that there is insufficient memory to store a Maximum Transmission Unit (MTU), it calls netif_stop_queue to stop the egress queue. This prevents further transmissions, as the kernel knows it will fail and waste resources.This responsibility lies with the driver.
+
+
+#### Scheduling for Transmission:
+
+```dev_queue_xmit``` dequeues a frame using one of two methods:
+
+1. Interface with Traffic Control (QoS layer): Through ```qdisc_run```.
+2. Directly pass the frame to the device’s hard_start_xmit.
+
+[more details of ```dev_queue_xmit```](<https://abcdxyzk.github.io/blog/2015/08/25/kernel-net-dev_queue_xmit/>)
+
+The sole parameter is a structure containing sk_buff:
+
+1. ```skb→dev``` specifies the outgoing device.
+2. ```skb→data``` points to the payload.
+
+When the egress queue is closed, netif_schedule is called to schedule the device:
+
+1. Adds the frame to the head of each CPU’s output queue.
+2. Schedules a softirq.
+
+#### Handling NET_TX_SOFTIRQ: ```net_tx_action```
+1. Handles tasks that can be deferred.
+2. When transmission is complete, dev_kfree_skb_irq is called to notify that the related buffer can be released.
+#### Traffic Control(TC)
+
```

Comment:
Is TBD item?

Reply:
I think that the topic of TC is less relevant to the rest of the article so I will remove this section